### ISSUE TITLE
doubts about your contract

### ISSUE THREAD

**Issue by damianlluch** [WillChomrak/Solidity-NFT-Minting-Contract#1]
Hi, I have read your contract, but it is not clear to me why you say it is a "lazy minting"....
The examples there are of lazy minting implementations are done with the encryption and decryption of the signature of the user who performs the "lazy minting".
Why do you say that what you did in the contract is lazy minting?

Greetings

**Comment by WillChomrak** [comment]
Hi Damian thanks for your comment. 

As I understand it, lazy minting refers to a process that does not mint all the NFTs on deployment, but allows them to each be minted at the time of purchase. This allows gas to be paid by the address that is minting one (or more) of the nfts, reducing the gas cost associated with deployment. The term does not refer to encryption of the user signature as far as I understand.

**Comment by distbit0** [comment]
So @WillChomrak could you confirm that I could use this code to create say, 10,000 NFTs, while only a few hundred dollars of gas on ethereum (to deploy this contract) and that the NFTs will be displayed on Opensea? 

If so, that is awesome, as usually batch minting 10k NFTs on ethereum costs several $100k. 
Do you have any Opensea collections you have made using this code that I can look at btw? Thanks! :)

**Comment by WillChomrak** [comment]
Hey @disbit! Back when I made this I tested to see approx what it would cost to deploy a 10,000 NFT collection and it was somewhere in the range of $500-$1000 CAD. I would imagine that is much lower right now because of the drop in ETH price.

I used this contract to deploy a 10,000 pfp NFT collection on Polygon (Solidity works perfectly on Polygon) and it cost only a couple dollars. The collection was a gag collection that had some sales but didn't go anywhere.  Check it out here if you'd like (not safe for work) http://nfpp.fun

**Comment by TheFunkySkullsNFT** [comment]
> So @WillChomrak could you confirm that I could use this code to create say, 10,000 NFTs, while only a few hundred dollars of gas on ethereum (to deploy this contract) and that the NFTs will be displayed on Opensea?
> 
> If so, that is awesome, as usually batch minting 10k NFTs on ethereum costs several $100k. Do you have any Opensea collections you have made using this code that I can look at btw? Thanks! :).


late to the party.

you both are wrong. that's not how lazy minting works at all. you would not see your NFTs on Opensea until they are minted by yourself or on your minting dapp or you upload each one by one to Opensea or with a script and use OpenSea erc1155 contract.

now, this contract is interesting because you have 2 "owners" splitting the profits
 but in reality the owner is the wallet that deploys the contract unless it transfers or renounce the contract.. but that's a different story.. and still interesting 

in reality you can not implement a lazy-minting function first because to my knowledge it doesn't exist. now, marketplaces use a erc1155 contract so they can mint on demand to one single contract (so, that means ALL lazy "minted" nft on opensea or any other marker belong to their respective erc1155 contract and that means millions and millions more of nfts are included in the same contract) and sure you can implement something like that in you own website using the erc721 contract, but you would have to code some script (js) to link all metadata files and images together so they would feel like they belong together and once some one buys them they would see them on their wallet... but you would still have to upload each one individually and set the metadata.. or use a script... so at the end they would be regular images with some under the hood script linking the metadada and then once someone buys them they become nfts.. yes it sounds complicated but its not so much. againg lazy-minting just meant... nice jepegs until someone pays to mint/create nfts...
the question is how opensea is able to display the lazy-minted images??? well they use a script to link all images and all  metadata together and ad it to a database and once someone mint one the nft gets created under the erc1155 contract..

on a side note. if you have couple extra hundred bucks you can deploy your own erc721 smart contract and mint all 10,000 nft and then they would show in Opensea.

or better yet get a partner and use this smart contract and split the profits 50-50 wink wink.


one last thing tho. 
 if once you deploy your contract and hit "transferToken" you're creating a transaction that cost gas.  you see the code below, there shows "external payable", and that's exactly what it means and you still need to mint the token. tokenTransfer its not free minting.

safeTransferFrom(address from, address to, uint256 tokenId)
external

Safely transfers tokenId token from from to to, checking first that contract recipients are aware of the ERC721 protocol to prevent tokens from being forever locked.

Requirements:

    from cannot be the zero address.

    to cannot be the zero address.

    tokenId token must exist and be owned by from. XXXX the doesn't exist until is minted XXXX

    If the caller is not from, it must be have been allowed to move this token by either [approve](https://docs.openzeppelin.com/contracts/3.x/api/token/erc721#IERC721-approve-address-uint256-) or [setApprovalForAll](https://docs.openzeppelin.com/contracts/3.x/api/token/erc721#IERC721-setApprovalForAll-address-bool-).

    If to refers to a smart contract, it must implement [IERC721Receiver.onERC721Received](https://docs.openzeppelin.com/contracts/3.x/api/token/erc721#IERC721Receiver-onERC721Received-address-address-uint256-bytes-), which is called upon a safe transfer.

Emits a [Transfer](https://docs.openzeppelin.com/contracts/3.x/api/token/erc721#IERC721-Transfer-address-address-uint256-) event

**Comment by WillChomrak** [comment]
Hey @TyWatson thanks for the reply. I'll try to address everything you mention.

Lazy minting does exist and that is what this contract does. It's super common in the NFT community from what I understand. The advantage is that by doing a lazy mint (minting only at the time of purchase) the buyer doesn't know what attributes they are getting in advance. This gives everyone the same chance to get a super-rare! Also the gas required to mint the collection is not all payed by the person who deploys the contract, but is spread out across each mint transaction.

I didn't look into it but it seems that Opensea simply identifies all the ERC721 tokens in a wallet and will display the image/details if there is metadata associated with a given token (usually at an IPFS address). 

This contract is a one of three parts that you can you use to generate, sell, and lazy-mint NFTs. Maybe checking out the other parts will help clarify some details.

Only read-only transactions are free on the Ethereum blockchain. Since transferring a token requires the contract write to the blockchain, it costs gas.

TLDR; Lazy-minting does exist and this contract works. If used correctly your NFTs will be minted at the time of purchase and Opensea will then display the NFT and metadata. 

Cheers!

**Comment by TheFunkySkullsNFT** [comment]
I didn't meant that lazy minting desn't exist, I meant that there's not a function nor method or construct that implement lazy minting...  minting its a name market places use to allow regular images to become nft by being minted at the time of purchase.
yeah the problem is with terminology. you're using the words "lazy minting" wrong but you're right most nft projects use this approach and its called the "reveal function".

if you go to open sea you can see the lazy minted nfts before purchasing. but in most projects as you said you can't see the nft until it get minted and that's why is call the "reveal". "sure you can set the reveal to a even later date than the mining day" but still thats the function call.
But again you're using the words lazy minting for wrong when you meant "reveal". nothing bad im just trying to correct you that's all.
btw I like it how you implemented the split profit method.

at the end of the day you cannot see nft of your project until they are minted and you wont be able to use the "_safeTransferFrom" function until the nft gets created/minted. yes even if you are the owner.

so again is just wrong terminology. but everything else its find. your contract looks fine and I'm sure people can use it
